Revamped server and client side n2f

shashwat-sh opened the pull request:
- Both server and client near-to-far transformations fully functional with the new refactored data structures
- All n2f functionality moved into the data subfolder
- N2f plugin deleted
- Associated back-end PR: https://github.com/flexcompute/tidy3d-core/pull/116

tylerflex commented on tidy3d/components/data/near2far_radiation_vectors.py:
```diff
@@ -485,345 +492,169 @@ def integrate_for_one_theta(i_th: int):
         """
 
+        freqs = np.atleast_1d(self.frequencies)
+        if monitor.plane_axis == 0:
```
see my earlier comments on unpop axis. some of this repeated code could also potentially be refactored into their own methods.

tylerflex commented on tidy3d/components/data/sim_data.py:
```diff
@@ -1,18 +1,21 @@
 from ..viz import equal_aspect, add_ax_if_none
 from ...log import log, DataError
 
+MonitorDataType = Union[MonitorDataType, Near2FarDataType]
```
see earlier comment, maybe this is better done in the `near2far_data.py` when `Near2FarDataType` is defined.

tylerflex commented on tidy3d/components/monitor.py:
```diff
@@ -447,6 +383,247 @@ def storage_size(self, num_cells: int, tmesh: int) -> int:
+        None,
+        title="Normal vector orientation",
+        description=":class:`.Direction` of the surface monitor's normal vector w.r.t. "
+        "the positive x, y or z unit vectors. Must be one of '+' or '-'.",
```
add note saying that this only applies to surfaces maybe?

tylerflex commented on tidy3d/components/monitor.py:
```diff
@@ -447,6 +383,247 @@ def storage_size(self, num_cells: int, tmesh: int) -> int:
+        units=MICROMETER,
+    )
+
+    def storage_size(self, num_cells: int, tmesh: ArrayLike[float, 1]) -> int:
```
Let's put the docstring
```py
"""Size of monitor storage given the number of points after discretization."""
```
under all of these `storage_size` methods as part of this PR.

tylerflex commented on tidy3d/components/monitor.py:
```diff
@@ -447,6 +383,247 @@ def storage_size(self, num_cells: int, tmesh: int) -> int:
+    ...     )
+    """
+
+    _data_type: Literal["Near2FarCartesianData"] = pydantic.Field("Near2FarCartesianData")
```
we dont need these as long as the monitor -> data mapping is in `DATA_TYPE_MAP`, can remove

tylerflex commented on tidy3d/components/data/near2far_radiation_vectors.py:
```diff
@@ -371,25 +365,30 @@ def _resample_surface_currents(
+        frequency: float,
+        theta: ArrayLikeN2F,
+        phi: ArrayLikeN2F,
+        surface: Near2FarSurface,
```
are `Near2FarSurface` required anymore? or can this just be a `Near2FarMonitorType`?

shashwat-sh replied:
`Near2FarSurface` wraps around the near-field `FieldMonitor` which stores the fields to be integrated, while `Near2FarMonitorType` stores information about the far-field observation grid, so I think both are necessary for client-side n2f computation. An alternative is for `Near2FarMonitorType` to store both near field surface currents and far field radiation vectors, but that would make the monitor not atomic?

tylerflex replied:
Hm, so where I'm still a bit unclear I guess is how the linking between the near field and far field takes place. For example, looking at `test_n2f_monitors`, it seems you add some near field monitors and some far field n2f monitors and it just knows how to handle it? Could you explain this a bit more? Specifically, how each Near2FarDataType "knows" which near field data to generate its data from and where `Near2FarSurface` comes in. I thought I understood earlier but now looking at it seems I don't quite get it.

shashwat-sh replied:
The two possible ways to compute far fields are:

1. `RadiationVectors` --> computes radiation vectors locally from `FieldMonitor`s --> returns the appropriate `Near2FarData`.
2. Solver uses `Near2FarMonitors` to return the appropriate `Near2FarData`.

In both cases, we end up with a `Near2FarData` containing radiation vectors. Its methods can then be used to compute far fields, RCS, etc. Also, in both cases we need to supply a `Near2FarMonitor` which specifies far field angles / points for which radiation vectors are computed.

tylerflex commented on tidy3d/components/data/near2far_radiation_vectors.py:
```diff
@@ -485,345 +492,169 @@ def integrate_for_one_theta(i_th: int):
+                    _N_th, _N_ph, _L_th, _L_ph = self._radiation_vectors_for_surface(
+                        frequency, theta, phi, surface, self.currents[surface.monitor.name]
+                    )
+                    N_theta[i, j, idx_f] += _N_th
```
I think for now this is ok, but there seems to be a pattern in here where the same operations are copied and pasted over a set of variables (`N_theta`, `N_phi`, `L_theta, `L_phi` in this case).  There's nothing wrong with this, but it would probably help to move towards using the data structures and comprehensions (where applicable) to reduce code redundancy and make things a bit more general.  For example 
```py
rad_vector_names = ('Ntheta', 'Nphi', 'Ltheta', 'Lphi')
rad_vectors = {name: np.zeros((len(ux), len(uy), len(freqs)), dtype=complex) for name in rad_vector_names}
...
for ...
    rad_vectors_surface = self._radiation_vectors_for_surface(...)
    for rad_vector, rad_vector_surface in zip(rad_vectors, rad_vectors_surface):
        rad_vector[i, j, idx_f] += rad_vector_surface

fields = {name:  Near2FarKSpaceDataArray(rad_vector, coords=coords) for name, rad_vector in rad_vectors.items()}
return Near2FarKSpaceData(monitor=monitor, **fields)
```

It would be a lot of work to refactor this, but I think it would be good to keep this in mind in the future if realizing that a lot of lines are getting copied and pasted.  Just keeps things more compact and maintainable.

tylerflex reviewed (changes requested):
great work @shashwat-sh. These new changes are going to be so useful. Had a lot of comments, but it is a pretty major PR so that's expected :).

shashwat-sh commented on tidy3d/components/grid/grid.py:
```diff
@@ -313,7 +313,10 @@ def _yee_h(self, axis: Axis):
         return Coords(**yee_coords)
 
-    def discretize_inds(self, box: Box, extend: bool = False) -> List[Tuple[int, int]]:
+    # pylint:disable=too-many-locals
```
Without it I do get the "too many locals" warning, so it does seem to work when placed before the def line?

shashwat-sh commented on tidy3d/components/data/near2far_radiation_vectors.py:
```diff
@@ -68,11 +73,11 @@ class Near2Far(Tidy3dBaseModel):
-        title="Frequency",
-        description="Frequency to select from each :class:`.Near2FarSurface` for projection.",
-        units=HERTZ,
+    resample: bool = pydantic.Field(
```
That's a good idea - I made `pts_per_wavelength` 10 by default to mimic the current behaviour, but made it so that it can be set to `None` to turn off resampling (also made this clear in the doc strings). The `resample` field is then no longer needed, which makes things simpler.

shashwat-sh commented on tidy3d/components/data/near2far_data.py:
```diff
@@ -0,0 +1,732 @@
+
+Near2FarDataTypes = (Near2FarAngleData, Near2FarCartesianData, Near2FarKSpaceData)
+Near2FarDataType = Union[Near2FarDataTypes]
+DATA_TYPE_MAP = DATA_TYPE_MAP.update(
```
I had it that way at the start but it gave errors in the backend test, e.g.:
```    sim_data = SimulationData(
File "pydantic/main.py", line 331, in pydantic.main.BaseModel.__init__
pydantic.error_wrappers.ValidationError: 3 validation errors for SimulationData
monitor_data -> n2f_angle
 No match for discriminator 'type' and value 'Near2FarAngleData' (allowed values: 'FieldData', 'FieldTimeData', 'PermittivityData', 'ModeSolverData', 'ModeData', 'FluxData', 'FluxTimeData') (type=value_error.discriminated_union.invalid_discriminator; discriminator_key=type; discriminator_value=Near2FarAngleData; allowed_values='FieldData', 'FieldTimeData', 'PermittivityData', 'ModeSolverData', 'ModeData', 'FluxData', 'FluxTimeData')```

I added `MonitorDataType = Union[MonitorDataType, Near2FarDataType]` below the line you marked, but maybe that's not the right way to do it?

tylerflex replied:
so these lines won't get run until the module is imported I believe.  So if someone imports `SimulationData` from `sim_data.py`, it will try importing `MonitorDataType` from `.monitor_data`. Maybe adding the `MonitorDataType = Union[MonitorDataType, Near2FarDataType]` to `near2far_data.py` never gets run because that module isn't imported.   I wonder if the best is to import everything into `data/__init__.py` and then construct the `MonitorDataType` and `DATA_TYPE_MAP` directly in there.  My worry is that if someone imports `MonitorDataType` from `monitor_data.py`, it might be different from importing it from `sim_data.py`. We should make sure it's consistent.

shashwat-sh replied:
Sorry, I'm not sure I understood the suggestion. Right now, in `near2far_data.py` we have

```python
class AbstractNear2FarData(MonitorData): # needs to import from monitor_data.py
...
class Near2FarAngleData(AbstractNear2FarData):
...
class Near2FarCartesianData(AbstractNear2FarData):
...
class Near2FarKSpaceData(AbstractNear2FarData):
...
Near2FarDataTypes = (Near2FarAngleData, Near2FarCartesianData, Near2FarKSpaceData)
Near2FarDataType = Union[Near2FarDataTypes]
```
In order to define the union `Near2FarDataType` in `monitor_data.py`, we'd need to somehow avoid the circular import.
If I interpret your suggestion correctly: we create a new file `near2far_utils.py` which has:
```python
from near2far_data import Near2FarDataTypes
Near2FarDataType = Union[Near2FarDataTypes]
```
then import this file into `monitor_data.py` and update `MonitorDataType` there? Isn't this still a circular import, just indirect?

Another option is to make a more general `utils.py` file in the `data` folder, containing
```python
from monitor_data import MonitorDataTypes
from near2far_data import Near2FarDataTypes

MonitorDataType = Union[MonitorDataTypes, Near2FarDataTypes]
```
and then in `sim_data.py` put `from utils import MonitorDataType`?

Or did you mean we should move the actual classes from `near2far_data.py` into `monitor_data.py`?

shashwat-sh commented:
Thanks for the comments @tylerflex ! I've made all the changes - the only unresolved issue is the place where `MonitorDataType` is defined (see my response to your comment above). It seems that redefining the union in `near2far_data.py` doesn't get registered?

tylerflex commented:
Dont forget to remove `sim_data.py.orig` too.

tylerflex reviewed (changes requested):
Thanks @shashwat-sh I think the most recent change (importing all the monitor types into `sim_data.py` and creating the DATA_TYPE_MAP there) makes a lot of sense.  The only remaining thing I wonder about is the file setup. It seems like `near2far_data` is an extension of `monitor_data` and `near2far_radiation_vector` is an extension of `data_array`.  Should we consider renaming these files to make this more clear? ie `monitor_data_n2f.py` and `data_array_n2f.py`? or we could even consider subdirectories for `monitor_data` and `data_array`. Any thoughts? Otherwise, from my point of view this all looks good and can approve after this.

shashwat-sh commented:
Thanks, I've renamed `near2far_data.py` to `monitor_data_n2f.py'. The radiation vectors file is not quite an extension of the data arrays; it just has the plugin functionality for client-side calculation of radiation vectors. I renamed it to `generate_n2f_data.py` because it returns subclasses of `AbstractNear2FarData` - maybe that's clearer?

As for folders, I don't have any strong opinion but it seems a bit of an overkill?

tylerflex commented:
Oh I see, so the n2f data arrays are defined in `data_array.py` but this `generate_n2f_data.py` just defines the `Near2FarSurface` and `RadiationVectors`. So these are more things the user could use to perform the near2far on the client side?  If there's a chance the user will be importing these objects from this file, maybe we should name it something concise like `near2far.py` so that they import `Near2FarSurface` (say) from
```py
from tidy3d.components.data.near2far import Near2FarSurface
```
Does this make sense with the use case?

shashwat-sh commented:
Thanks @tylerflex , I've changed the name. Yes, the classes in that file are already in the __init__ files, so the user can do `td.Near2FarSurface` or any of the alternatives you suggested.
